# snpArcher

<img src="./docs/img/logo.png" alt="snpArcher logo" height="300"/>

snpArcher is a reproducible workflow optimized for nonmodel organisms and comparisons across datasets, built on the [Snakemake](https://snakemake.readthedocs.io/en/stable/index.html#) workflow management system. It provides a streamlined approach to dataset acquisition, variant calling, quality control, and downstream analysis.

### Usage

For usage instructions and complete documentation, please visit our [docs](https://snparcher.readthedocs.io/en/latest/).

### Datasets generated by snpArcher

A number of resequencing datasets have been run with snpArcher generating consistent variant calls, available via [Globus](https://www.globus.org/) in the [Comparative Population Genomics Data collection](https://app.globus.org/file-manager?origin_id=a6580c44-09fd-11ee-be16-195c41bc0be4&origin_path=%2F). Details of data processing are described [in our manuscript](https://www.biorxiv.org/content/10.1101/2023.06.22.546168v1). If you use any of these datasets in your projects, please cite both the [snpArcher paper](https://www.biorxiv.org/content/10.1101/2023.06.22.546168v1) and the original data producers.

# Modules

## *svArcher*: A Structural Variant Calling Extension module for snpArcher.

<div align="center">
<img src="workflow/modules/svArcher/img/svArcher_Logo.png" alt="svArcher Logo" width="150" height="190">
</div>

*[svArcher](workflow/modules/svArcher)* is a reproducible subworkflow designed to call structural variants (SVs) from short read genomic sequencing data. The pipeline utilizes an ensemble of structural variant callers (DELLY, Lumpy and Wham) that call SV's independently. Results are merged and filtered with SURVIVOR. Current parameters (1000 3 1 1 0 50) are set to achieve a concensus of SV's from all 3 callers per sample.

## Usage

The *svArcher* module is deactivated by default. Set svCall option to `True` in your workflow `config.yaml` file.

Additionally, provide a `.csv` file with a list of contigs/chromosome ID's present in the reference sequence file. Each ID must start on its own line.

```yaml
svCall: False # Set to true to activate svArcher
include_contigs: "/path/to/ref_contigs.csv"
```

## Benchmarking

This module enables users to generate benchmark comparison metrics for deletion, duplication and inversion structural variant types. To activate this feature, set svBenchmark to `True`.

We recommend creating a single directory to store all ground truth `.vcf` files. When configuring options below, provide the absolute path to this directory.

```yaml
svBenchmark: False
# Benchmark Truth set paths
deletions: "benchmark_truth/"
duplications: "benchmark_truth/"
inversions: "benchmark_truth/"

```

### Citing snpArcher

- Cade D Mirchandani, Allison J Shultz, Gregg W C Thomas, Sara J Smith, Mara Baylis, Brian Arnold, Russ Corbett-Detig, Erik Enbody, Timothy B Sackton, A fast, reproducible, high-throughput variant calling workflow for population genomics, Molecular Biology and Evolution, 2023;, msad270, https://doi.org/10.1093/molbev/msad270
- Also, make sure to cite the tools you used within snpArcher.
